torch/jit.py

```diff
@@ -1,26 +1,375 @@
+        out, extras = _unflatten(out_vars, out_struct)
+        assert len(extras) == 0
+        return out
+
+    def has_trace_for(self, *args):
+        in_vars, in_struct, is_volatile, in_key = self._process_args(args)
+        ktrace = self.ktrace_cache.get(in_key)
+        if ktrace is None:
+            return False
+        return ktrace.maybe_closure() is not None
+
+    # TODO: Provide more compiled code management utility methods
+
+
+# CompiledModule memoizes multiple traces and switches between them based on
+# inputs provided to a call; a TraceForKey represents one such trace. Things
+# that need to be considered include non-Variable argument (e.g. num_layers=3;
+# compared by equality) or Variable flags and sizes. TraceForKey is the object
+# that is used to hold a trace / compiled code for a single input configuration
+# aka in_key.
+class TraceForKey(object):
+    # Lifecycle:
+    #   - We accumulate 'traces'
+    #   - At some point, one of these traces becomes complete ('is_complete'
+    #     is True).  This occurs when we run enough backwards on a trace
+    #     to complete it (i.e., this is an external event to this object).
+    #   - Whenever we want to run this trace, we call 'maybe_closure'.  This
+    #     returns None if we don't have a complete trace yet, or the
+    #     autograd closure to actually run the trace if we do.
+    def __init__(self, f, nderivs=1, optimize=True, volatile=False):
+        self.f = f
+        # TODO: Not convinced about this volatile special case...
+        self.nderivs = nderivs if not volatile else 0
+        self.optimize = optimize
+        self.traces = []
+        self.closure = None
+        self.out_struct = None # initialized when we call trace, checked thereafter
+
+    # The signature here is a little goofy; it's a perf optimization
+    def add_trace(self, args, in_vars, in_struct):
+        # TODO: Deduplicate this code
+        @raw_trace(nderivs=self.nderivs)
+        def traced_f(in_vars, in_struct):
+            #args, _params = _unflatten(in_vars, in_struct)
+            return _flatten(self.f(*args))
+
+        trace, (out_vars, out_struct) = traced_f(in_vars, in_struct)
+        if self.out_struct is None:
+            self.out_struct = out_struct
+        else:
+            # TODO: in debug mode, assert the output structs are same
+            pass
+        self.traces.append(trace)
+        return out_vars, out_struct
+
+    def maybe_closure(self):
+        if self.closure:
+            return self.closure
+
+        # GC expired traces
+        self.traces = [t for t in self.traces if not t.is_expired]
+
+        # Search for a complete trace
+        complete_trace = None
+        for trace in self.traces:
+            if trace.is_complete:
+                complete_trace = trace
+                self.traces = []
+
+        if complete_trace is None:
+            return None
+
+        def _run_pass(p, trace):
+            # TODO: dump the trace with names
```

Comment:
This patch can't go in until all the debug logic that was in the previous version is restored. I can't optimize without it.